File: backend/dotnet/src/Mateu.Uidl/ImportWizard.cs
using System.ComponentModel.DataAnnotations;
using System.Globalization;
using System.Reflection;
using System.Text;

namespace Mateu.Uidl;

/// <summary>One row of the import wizard's mapping grid: a CSV column, a sample value from the
/// first data row, and the target row-class field it pours into (TargetField is the only editable
/// cell — an inline select whose options are the row class's assignable field names; blank = skip
/// the column). (C# analogue of Java's ColumnMapping.)</summary>
public class ColumnMapping
{
    [ReadOnly, Label("CSV column")] public string? CsvColumn { get; set; }
    [ReadOnly, Label("Sample")] public string? Sample { get; set; }
    [Label("Target field")] public string? TargetField { get; set; }
}

/// <summary>One problem found while validating a CSV data row in the import wizard.
/// (C# analogue of Java's RowIssue.)</summary>
public class RowIssue
{
    [Label("Line")] public int Line { get; set; }
    [Label("Column")] public string? Column { get; set; }
    [Label("Value")] public string? Value { get; set; }
    [Label("Problem")] public string? Problem { get; set; }
}

/// <summary>
/// Guided CSV import wizard (the C# analogue of Java's <c>ImportWizard&lt;Row&gt;</c> archetype):
/// upload (or paste) a CSV, map its columns onto the row class's properties (auto-mapped by name
/// similarity, adjustable in an inline-editable grid whose target-field cell is a select fed by
/// the wizard's <see cref="IOptionsSupplier"/>), review a validation report (conversion failures
/// + DataAnnotations violations per line), then import the valid rows through
/// <see cref="ImportRows"/>. The result step shows imported/skipped counts.
///
/// <para>Steps: 1 upload → 2 mapping → 3 validation → 4 result. Moving forward computes the next
/// step's content (<see cref="OnNext"/>); the import itself runs on the validation step's Next —
/// the port's wizards have no completion-action button, Next/Finish is the driver.</para>
///
/// <para>Extend it, route it with <c>[UI]</c>, and implement <c>ImportRows</c>:</para>
/// <code>
/// [UI("product-import")]
/// public class ProductImport : ImportWizard&lt;Product&gt;
/// {
///     protected override void ImportRows(List&lt;Product&gt; rows) => rows.ForEach(repository.Save);
/// }
/// </code>
/// </summary>
public abstract class ImportWizard<Row> : Wizard, IOptionsSupplier where Row : new()
{
    [Step(1), FileUpload(".csv"), Label("CSV file")]
    public string? File { get; set; }

    [Step(1), Stereotype("textarea"), Label("...or paste the CSV")]
    public string? Pasted { get; set; }

    [Step(2), InlineEditing, Label("Column mapping")]
    public List<ColumnMapping> Mappings { get; set; } = [];

    [Step(3), PlainText, Label("Valid rows")]
    public int ValidRows { get; set; }

    [Step(3), PlainText, Label("Rows with problems")]
    public int InvalidRows { get; set; }

    [Step(3), ReadOnly, Label("Issues")]
    public List<RowIssue> Issues { get; set; } = [];

    [Step(4), PlainText, Label("Imported")]
    public int Imported { get; set; }

    [Step(4), PlainText, Label("Skipped")]
    public int Skipped { get; set; }

    /// <summary>Receives the valid typed rows when the user confirms the import (Next on the
    /// validation step).</summary>
    protected abstract void ImportRows(List<Row> rows);

    public override Message Complete() => new($"Imported {Imported} rows ({Skipped} skipped)");

    public override void OnNext(int from, int to)
    {
        if (from == 1 && to == 2) PopulateMappings();
        if (from == 2 && to == 3) ValidateRows();
        if (from == 3 && to == 4)
        {
            var (valid, _, invalidCount) = Assemble();
            ImportRows(valid);
            Imported = valid.Count;
            Skipped = invalidCount;
        }
    }

    /// <summary>The mapping grid's target-field select options: "— skip —" plus the row class's
    /// assignable (coercible, writable) property names.</summary>
    public IReadOnlyList<Option> Options(string fieldName)
    {
        if (fieldName != "targetField") return [];
        var options = new List<Option> { new("", "— skip —") };
        options.AddRange(AssignableProperties().Select(p => new Option(CamelCase(p.Name), CamelCase(p.Name))));
        return options;
    }

    /// <summary>Parses the CSV header + first data row and auto-maps columns by name similarity
    /// (exact case-insensitive match first, then containment either way; blank = unmapped).</summary>
    private void PopulateMappings()
    {
        Mappings = [];
        var records = CsvParser.Parse(Content());
        if (records.Count == 0) return;
        var header = records[0];
        var sample = records.Count > 1 ? records[1] : [];
        var targets = AssignableProperties().Select(p => CamelCase(p.Name)).ToList();
        for (var i = 0; i < header.Count; i++)
            Mappings.Add(new ColumnMapping
            {
                CsvColumn = header[i],
                Sample = i < sample.Count ? sample[i] : "",
                TargetField = AutoMap(header[i], targets),
            });
    }

    private static string AutoMap(string column, List<string> targets)
    {
        var normalized = column.Trim().ToLowerInvariant();
        foreach (var target in targets)
            if (target.ToLowerInvariant() == normalized)
                return target;
        foreach (var target in targets)
        {
            var name = target.ToLowerInvariant();
            if (normalized.Length > 0 && (name.Contains(normalized) || normalized.Contains(name)))
                return target;
        }
        return "";
    }

    /// <summary>Re-parses and validates every data row when entering the validation step.</summary>
    private void ValidateRows()
    {
        var (valid, issues, invalidCount) = Assemble();
        ValidRows = valid.Count;
        InvalidRows = invalidCount;
        Issues = issues;
    }

    /// <summary>Builds typed rows from the parsed CSV per the mappings: coerces each cell into the
    /// target property (string/numerics/bool/dates/enum), collects conversion failures, then runs
    /// the standard DataAnnotations validation ([Required], [Range], …) over the built row — the
    /// port's server-side validation surface.</summary>
    private (List<Row> Valid, List<RowIssue> Issues, int InvalidCount) Assemble()
    {
        var valid = new List<Row>();
        var issues = new List<RowIssue>();
        var invalidCount = 0;
        var records = CsvParser.Parse(Content());
        if (records.Count < 2 || Mappings.Count == 0) return (valid, issues, invalidCount);
        var header = records[0];
        for (var r = 1; r < records.Count; r++)
        {
            var line = r + 1; // 1-based, counting the header line
            var record = records[r];
            var rowIssues = new List<RowIssue>();
            var row = new Row();
            foreach (var mapping in Mappings)
            {
                if (string.IsNullOrWhiteSpace(mapping.TargetField)) continue;
                var index = header.IndexOf(mapping.CsvColumn ?? "");
                if (index < 0 || index >= record.Count) continue;
                var raw = record[index];
                if (string.IsNullOrWhiteSpace(raw)) continue; // requiredness → the validation pass
                var property = AssignableProperties()
                    .FirstOrDefault(p => CamelCase(p.Name) == mapping.TargetField);
                if (property is null)
                {
                    rowIssues.Add(new RowIssue
                    {
                        Line = line, Column = mapping.CsvColumn, Value = mapping.TargetField,
                        Problem = "Unknown target field",
                    });
                    continue;
                }
                try
                {
                    property.SetValue(row, Coerce(property.PropertyType, raw.Trim()));
                }
                catch
                {
                    rowIssues.Add(new RowIssue
                    {
                        Line = line, Column = mapping.CsvColumn, Value = raw,
                        Problem = "Cannot convert to " + (Nullable.GetUnderlyingType(property.PropertyType)
                                                          ?? property.PropertyType).Name,
                    });
                }
            }
            rowIssues.AddRange(ValidateConstraints(row, line));
            if (rowIssues.Count == 0)
            {
                valid.Add(row);
            }
            else
            {
                invalidCount++;
                issues.AddRange(rowIssues);
            }
        }
        return (valid, issues, invalidCount);
    }

    /// <summary>DataAnnotations validation over the built row ([Required], [Range], [MinLength]…),
    /// reported per line with the CSV column mapped onto the failing member.</summary>
    private List<RowIssue> ValidateConstraints(Row row, int line)
    {
        var results = new List<ValidationResult>();
        Validator.TryValidateObject(row!, new ValidationContext(row!), results, validateAllProperties: true);
        return results.Select(result =>
        {
            var member = result.MemberNames.FirstOrDefault() ?? "";
            var fieldId = CamelCase(member);
            var value = typeof(Row).GetProperty(member)?.GetValue(row);
            return new RowIssue
            {
                Line = line,
                Column = Mappings.FirstOrDefault(m => m.TargetField == fieldId)?.CsvColumn ?? fieldId,
                Value = value?.ToString() ?? "",
                Problem = result.ErrorMessage ?? "Invalid value",
            };
        }).ToList();
    }

    /// <summary>The row-class properties a CSV column can pour into (writable basics, enums and
    /// temporals).</summary>
    private static List<PropertyInfo> AssignableProperties() =>
        typeof(Row).GetProperties(BindingFlags.Public | BindingFlags.Instance)
            .Where(p => p is { CanRead: true, CanWrite: true } && IsCoercible(p.PropertyType))
            .ToList();

    private static bool IsCoercible(Type type)
    {
        type = Nullable.GetUnderlyingType(type) ?? type;
        return type == typeof(string) || type == typeof(bool)
            || type == typeof(byte) || type == typeof(short) || type == typeof(int) || type == typeof(long)
            || type == typeof(float) || type == typeof(double) || type == typeof(decimal)
            || type == typeof(DateOnly) || type == typeof(DateTime) || type.IsEnum;
    }

    private static object Coerce(Type type, string raw)
    {
        type = Nullable.GetUnderlyingType(type) ?? type;
        if (type == typeof(string)) return raw;
        if (type == typeof(bool)) return bool.Parse(raw);
        if (type == typeof(byte)) return byte.Parse(raw, CultureInfo.InvariantCulture);
        if (type == typeof(short)) return short.Parse(raw, CultureInfo.InvariantCulture);
        if (type == typeof(int)) return int.Parse(raw, CultureInfo.InvariantCulture);
        if (type == typeof(long)) return long.Parse(raw, CultureInfo.InvariantCulture);
        if (type == typeof(float)) return float.Parse(raw, CultureInfo.InvariantCulture);
        if (type == typeof(double)) return double.Parse(raw, CultureInfo.InvariantCulture);
        if (type == typeof(decimal)) return decimal.Parse(raw, CultureInfo.InvariantCulture);
        if (type == typeof(DateOnly)) return DateOnly.Parse(raw, CultureInfo.InvariantCulture);
        if (type == typeof(DateTime)) return DateTime.Parse(raw, CultureInfo.InvariantCulture);
        if (type.IsEnum) return Enum.Parse(type, raw, ignoreCase: true);
        throw new NotSupportedException(type.Name);
    }

    /// <summary>The CSV text: the uploaded file (base64-decoded data URI) when present, else the
    /// pasted text.</summary>
    private string Content()
    {
        if (!string.IsNullOrWhiteSpace(File)) return DecodeDataUri(File!);
        return Pasted ?? "";
    }

    /// <summary>Decodes a data: URI into text (base64 payload after the first comma, UTF-8);
    /// a plain (non-data-URI) value is returned as-is.</summary>
    private static string DecodeDataUri(string value)
    {
        if (!value.StartsWith("data:")) return value;
        var comma = value.IndexOf(',');
        if (comma < 0) return "";
        var meta = value[..comma];
        var payload = value[(comma + 1)..];
        if (meta.Contains(";base64"))
        {
            try
            {
                return Encoding.UTF8.GetString(Convert.FromBase64String(payload));
            }
            catch (FormatException)
            {
                return "";
            }
        }
        return Uri.UnescapeDataString(payload);
    }

    private static string CamelCase(string name) =>
        name.Length == 0 ? name : char.ToLowerInvariant(name[0]) + name[1..];
}

/// <summary>Minimal RFC-4180-ish CSV parser for the import wizard: quoted fields may embed the
/// separator, doubled quotes ("") and newlines; rows end at \n, \r\n or \r. Both , and ; separate
/// fields — the separator is detected by counting the candidates (outside quotes) on the first
/// line. (C# port of Java's CsvParser.)</summary>
public static class CsvParser
{
    /// <summary>Parses the whole content into rows of cells; fully-blank lines are dropped.</summary>
    public static List<List<string>> Parse(string? content)
    {
        if (string.IsNullOrWhiteSpace(content)) return [];
        var separator = DetectSeparator(content);
        var rows = new List<List<string>>();
        var row = new List<string>();
        var cell = new StringBuilder();
        var inQuotes = false;
        var i = 0;
        var length = content.Length;
        while (i < length)
        {
            var c = content[i];
            if (inQuotes)
            {
                if (c == '"')
                {
                    if (i + 1 < length && content[i + 1] == '"') { cell.Append('"'); i += 2; continue; }
                    inQuotes = false;
                    i++;
                    continue;
                }
                cell.Append(c);
                i++;
                continue;
            }
            if (c == '"') { inQuotes = true; i++; continue; }
            if (c == separator) { row.Add(cell.ToString()); cell.Clear(); i++; continue; }
            if (c is '\r' or '\n')
            {
                row.Add(cell.ToString());
                cell.Clear();
                rows.Add(row);
                row = [];
                if (c == '\r' && i + 1 < length && content[i + 1] == '\n') i++;
                i++;
                continue;
            }
            cell.Append(c);
            i++;
        }
        if (cell.Length > 0 || row.Count > 0)
        {
            row.Add(cell.ToString());
            rows.Add(row);
        }
        rows.RemoveAll(cells => cells.All(string.IsNullOrWhiteSpace));
        return rows;
    }

    /// <summary>The field separator, detected on the first line: ; when it outnumbers , outside
    /// quotes, else ,.</summary>
    public static char DetectSeparator(string content)
    {
        var commas = 0;
        var semicolons = 0;
        var inQuotes = false;
        foreach (var c in content)
        {
            if (c == '"') inQuotes = !inQuotes;
            else if (!inQuotes)
            {
                if (c is '\r' or '\n') break;
                if (c == ',') commas++;
                if (c == ';') semicolons++;
            }
        }
        return semicolons > commas ? ';' : ',';
    }
}

File: backend/python/mateu_core/importwizard.py
"""Guided CSV import wizard — the Python analogue of Java's ``ImportWizard<Row>`` archetype.

Upload (or paste) a CSV, map its columns onto the row class's fields (auto-mapped by name
similarity, adjustable in an inline-editable grid whose target-field cell is a select fed by the
wizard's ``options()`` supplier), review a validation report (conversion failures + ``Required()``
violations per line), then import the valid rows through :meth:`ImportWizard.import_rows`. The
result step shows imported/skipped counts.

Steps: 1 upload → 2 mapping → 3 validation → 4 result. Moving forward computes the next step's
content (:meth:`ImportWizard.on_next`); the import itself runs on the validation step's Next —
the port's wizards have no completion-action button, Next/Finish is the driver.
"""

from __future__ import annotations

import base64
import binascii
import urllib.parse
from datetime import date, datetime
from decimal import Decimal
from enum import Enum
from typing import Annotated, Any, Generic, TypeVar, get_args, get_origin

from mateu_uidl import (
    FileUpload,
    InlineEditing,
    Label,
    Message,
    PlainText,
    ReadOnly,
    Required,
    Step,
    Stereotype,
    Wizard,
)

from .naming import camel_case
from .reflection import view_fields

Row = TypeVar("Row")


class ColumnMapping:
    """One row of the import wizard's mapping grid: a CSV column, a sample value from the first
    data row, and the target row-class field it pours into (``target_field`` is the only editable
    cell — an inline select whose options are the row class's assignable field names; blank =
    skip the column)."""

    csv_column: Annotated[str | None, ReadOnly(), Label("CSV column")] = None
    sample: Annotated[str | None, ReadOnly(), Label("Sample")] = None
    target_field: Annotated[str | None, Label("Target field")] = None

    def __init__(self, csv_column: str | None = None, sample: str | None = None,
                 target_field: str | None = None):
        self.csv_column = csv_column
        self.sample = sample
        self.target_field = target_field


class RowIssue:
    """One problem found while validating a CSV data row in the import wizard."""

    line: Annotated[int, Label("Line")] = 0
    column: Annotated[str | None, Label("Column")] = None
    value: Annotated[str | None, Label("Value")] = None
    problem: Annotated[str | None, Label("Problem")] = None

    def __init__(self, line: int = 0, column: str | None = None, value: str | None = None,
                 problem: str | None = None):
        self.line = line
        self.column = column
        self.value = value
        self.problem = problem


class ImportWizard(Wizard, Generic[Row]):
    """Extend it, route it with ``@ui``, and implement :meth:`import_rows`::

        @ui("product-import")
        class ProductImport(ImportWizard[Product]):
            def import_rows(self, rows):
                for row in rows:
                    repository.save(row)

    ``row_class()`` defaults to the ``Row`` generic argument."""

    file: Annotated[str | None, Step(1), FileUpload(accept=".csv"), Label("CSV file")] = None
    pasted: Annotated[str | None, Step(1), Stereotype("textarea"),
                      Label("...or paste the CSV")] = None
    mappings: Annotated[list[ColumnMapping], Step(2), InlineEditing(),
                        Label("Column mapping")] = ()
    valid_rows: Annotated[int, Step(3), PlainText(), Label("Valid rows")] = 0
    invalid_rows: Annotated[int, Step(3), PlainText(), Label("Rows with problems")] = 0
    issues: Annotated[list[RowIssue], Step(3), ReadOnly(), Label("Issues")] = ()
    imported: Annotated[int, Step(4), PlainText(), Label("Imported")] = 0
    skipped: Annotated[int, Step(4), PlainText(), Label("Skipped")] = 0

    # ── the developer's surface ─────────────────────────────────────────────────
    def import_rows(self, rows: list[Row]) -> None:
        """Receives the valid typed rows when the user confirms the import (Next on the
        validation step)."""
        raise NotImplementedError

    def row_class(self) -> type:
        """The class each CSV data row hydrates into; defaults to the ``Row`` generic argument."""
        for klass in type(self).__mro__:
            for base in getattr(klass, "__orig_bases__", ()):
                origin = get_origin(base)
                if isinstance(origin, type) and issubclass(origin, ImportWizard):
                    args = get_args(base)
                    if args and isinstance(args[0], type):
                        return args[0]
        raise TypeError(f"{type(self).__name__} must subscribe ImportWizard[Row]")

    # ── wizard machinery ────────────────────────────────────────────────────────
    def complete(self) -> Message:
        return Message(f"Imported {self.imported} rows ({self.skipped} skipped)")

    def on_next(self, from_step: int, to_step: int) -> None:
        if from_step == 1 and to_step == 2:
            self._populate_mappings()
        if from_step == 2 and to_step == 3:
            valid, issues, invalid_count = self._assemble()
            self.valid_rows = len(valid)
            self.invalid_rows = invalid_count
            self.issues = issues
        if from_step == 3 and to_step == 4:
            valid, _, invalid_count = self._assemble()
            self.import_rows(valid)
            self.imported = len(valid)
            self.skipped = invalid_count

    def options(self, field_name: str):
        """The mapping grid's target-field select options: "— skip —" plus the row class's
        assignable (coercible) field names."""
        if field_name != "targetField":
            return []
        options: list[tuple[str, str]] = [("", "— skip —")]
        options.extend(
            (camel_case(f.name), camel_case(f.name)) for f in _assignable_fields(self.row_class())
        )
        return options

    # ── internals ───────────────────────────────────────────────────────────────
    def _populate_mappings(self) -> None:
        """Parses the CSV header + first data row and auto-maps columns by name similarity
        (exact case-insensitive match first, then containment either way; blank = unmapped)."""
        self.mappings = []
        records = parse_csv(self._content())
        if not records:
            return
        header = records[0]
        sample = records[1] if len(records) > 1 else []
        targets = [camel_case(f.name) for f in _assignable_fields(self.row_class())]
        for i, column in enumerate(header):
            self.mappings.append(ColumnMapping(
                csv_column=column,
                sample=sample[i] if i < len(sample) else "",
                target_field=_auto_map(column, targets),
            ))

    def _assemble(self) -> tuple[list[Row], list[RowIssue], int]:
        """Builds typed rows from the parsed CSV per the mappings: coerces each cell into the
        target field (str/int/float/Decimal/bool/dates/Enum), collects conversion failures, then
        checks ``Required()`` fields — the port's server-side validation surface."""
        row_class = self.row_class()
        valid: list[Row] = []
        issues: list[RowIssue] = []
        invalid_count = 0
        records = parse_csv(self._content())
        mappings = list(self.mappings or [])
        if len(records) < 2 or not mappings:
            return valid, issues, invalid_count
        header = records[0]
        fields = {camel_case(f.name): f for f in _assignable_fields(row_class)}
        for r in range(1, len(records)):
            line = r + 1  # 1-based, counting the header line
            record = records[r]
            row_issues: list[RowIssue] = []
            row = row_class()
            for mapping in mappings:
                target = mapping.target_field
                if not target or not target.strip():
                    continue
                index = header.index(mapping.csv_column) if mapping.csv_column in header else -1
                if index < 0 or index >= len(record):
                    continue
                raw = record[index]
                if raw is None or not raw.strip():
                    continue  # leave the field default; requiredness → the constraint pass
                f = fields.get(target)
                if f is None:
                    row_issues.append(RowIssue(line, mapping.csv_column, target,
                                               "Unknown target field"))
                    continue
                try:
                    setattr(row, f.name, _coerce(f.type, raw.strip()))
                except (ValueError, KeyError, TypeError):
                    row_issues.append(RowIssue(
                        line, mapping.csv_column, raw,
                        "Cannot convert to " + getattr(f.type, "__name__", str(f.type)),
                    ))
            row_issues.extend(_validate_required(row_class, row, line, mappings))
            if not row_issues:
                valid.append(row)
            else:
                invalid_count += 1
                issues.extend(row_issues)
        return valid, issues, invalid_count

    def _content(self) -> str:
        """The CSV text: the uploaded file (base64-decoded data URI) when present, else the
        pasted text."""
        if self.file and self.file.strip():
            return _decode_data_uri(self.file)
        return self.pasted or ""


def _auto_map(column: str, targets: list[str]) -> str:
    normalized = column.strip().lower()
    for target in targets:
        if target.lower() == normalized:
            return target
    for target in targets:
        name = target.lower()
        if normalized and (normalized in name or name in normalized):
            return target
    return ""


def _assignable_fields(row_class: type):
    """The row-class fields a CSV column can pour into (coercible basics, enums and temporals)."""
    return [f for f in view_fields(row_class) if _is_coercible(f.type)]


def _is_coercible(t) -> bool:
    return t in (str, bool, int, float, Decimal, date, datetime) or (
        isinstance(t, type) and issubclass(t, Enum)
    )


def _coerce(t, raw: str):
    """A CSV cell into the target field type; raises on unconvertible values."""
    if t is str:
        return raw
    if t is bool:
        lowered = raw.lower()
        if lowered in ("true", "1", "yes"):
            return True
        if lowered in ("false", "0", "no"):
            return False
        raise ValueError(raw)
    if t is int:
        return int(raw)
    if t is float:
        return float(raw)
    if t is Decimal:
        return Decimal(raw)
    if t is date:
        return date.fromisoformat(raw)
    if t is datetime:
        return datetime.fromisoformat(raw)
    if isinstance(t, type) and issubclass(t, Enum):
        try:
            return t[raw]
        except KeyError:
            return t(raw)
    raise ValueError(raw)


def _validate_required(row_class: type, row, line: int, mappings) -> list[RowIssue]:
    """``Required()`` fields must end up non-None and non-blank — the Python port's validation
    surface (it has no Min/Max markers)."""
    issues: list[RowIssue] = []
    for f in _assignable_fields(row_class):
        if not f.has(Required):
            continue
        value = getattr(row, f.name, None)
        if value is None or (isinstance(value, str) and not value.strip()):
            field_id = camel_case(f.name)
            column = next((m.csv_column for m in mappings if m.target_field == field_id), field_id)
            issues.append(RowIssue(line, column, "", "Must not be empty"))
    return issues


def _decode_data_uri(value: str) -> str:
    """Decodes a ``data:`` URI into text (base64 payload after the first comma, UTF-8); a plain
    (non-data-URI) value is returned as-is."""
    if not value.startswith("data:"):
        return value
    comma = value.find(",")
    if comma < 0:
        return ""
    meta, payload = value[:comma], value[comma + 1:]
    if ";base64" in meta:
        try:
            return base64.b64decode(payload).decode("utf-8")
        except (binascii.Error, ValueError):
            return ""
    return urllib.parse.unquote(payload)


def parse_csv(content: str | None) -> list[list[str]]:
    """Minimal RFC-4180-ish CSV parser (the Python port of Java's CsvParser): quoted fields may
    embed the separator, doubled quotes ("") and newlines; rows end at \\n, \\r\\n or \\r. Both
    ``,`` and ``;`` separate fields — the separator is detected by counting the candidates
    (outside quotes) on the first line. Fully-blank lines are dropped."""
    if content is None or not content.strip():
        return []
    separator = detect_separator(content)
    rows: list[list[str]] = []
    row: list[str] = []
    cell: list[str] = []
    in_quotes = False
    i = 0
    length = len(content)
    while i < length:
        c = content[i]
        if in_quotes:
            if c == '"':
                if i + 1 < length and content[i + 1] == '"':
                    cell.append('"')
                    i += 2
                    continue
                in_quotes = False
                i += 1
                continue
            cell.append(c)
            i += 1
            continue
        if c == '"':
            in_quotes = True
            i += 1
            continue
        if c == separator:
            row.append("".join(cell))
            cell = []
            i += 1
            continue
        if c in ("\r", "\n"):
            row.append("".join(cell))
            cell = []
            rows.append(row)
            row = []
            if c == "\r" and i + 1 < length and content[i + 1] == "\n":
                i += 1
            i += 1
            continue
        cell.append(c)
        i += 1
    if cell or row:
        row.append("".join(cell))
        rows.append(row)
    return [cells for cells in rows if any(cell.strip() for cell in cells)]


def detect_separator(content: str) -> str:
    """The field separator, detected on the first line: ``;`` when it outnumbers ``,`` outside
    quotes, else ``,``."""
    commas = 0
    semicolons = 0
    in_quotes = False
    for c in content:
        if c == '"':
            in_quotes = not in_quotes
        elif not in_quotes:
            if c in ("\r", "\n"):
                break
            if c == ",":
                commas += 1
            if c == ";":
                semicolons += 1
    return ";" if semicolons > commas else ","
